File: codegen/results/v002/gemma4_31b_nothink/20260414_073749/go/007_beatles_interview/solution.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

type Relative struct {
	FirstName    string `json:"FirstName"`
	LastName     string `json:"LastName"`
	Relationship string `json:"Relationship"`
}

type Person struct {
	FirstName string     `json:"FirstName"`
	LastName  string     `json:"LastName"`
	Birthday  string     `json:"Birthday"`
	Age       int        `json:"Age"`
	Relatives []Relative `json:"Relatives"`
}

func calculateAge(birthdayStr string, referenceDate time.Time) int {
	// Input format from CSV seems to be M/D/YYYY (e.g., 10/9/1940)
	t, err := time.Parse("1/2/2006", birthdayStr)
	if err != nil {
		return 0
	}

	age := referenceDate.Year() - t.Year()
	// Check if birthday has occurred yet in the reference year
	if referenceDate.YearDay() < t.YearDay() {
		// This is a simplification; for production we'd handle leap years more precisely
		// but for standard age calculation:
		if referenceDate.Month() < t.Month() || (referenceDate.Month() == t.Month() && referenceDate.Day() < t.Day()) {
			age--
		}
	}
	return age
}

func parseName(fullName string) (string, string) {
	if fullName == "null" || fullName == "" {
		return "", ""
	}
	parts := strings.Fields(fullName)
	if len(parts) == 0 {
		return "", ""
	}
	first := parts[0]
	last := ""
	if len(parts) > 1 {
		last = parts[len(parts)-1]
	}
	return first, last
}

func main() {
	file, err := os.Open("input/input.csv")
	if err != nil {
		return
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return
	}

	if len(records) < 1 {
		fmt.Println("[]")
		return
	}

	// Reference date: July 1, 2025
	refDate := time.Date(2025, time.July, 1, 0, 0, 0, 0, time.UTC)
	
	var people []Person

	// Skip header row
	for i := 1; i < len(records); i++ {
		row := records[i]
		if len(row) < 7 {
			continue
		}

		// Parse Main Person
		fullPersonName := row[0]
		fName, lName := parseName(fullPersonName)
		
		// Format Birthday to YYYY-MM-DD
		bDayRaw := row[1]
		t, _ := time.Parse("1/2/2006", bDayRaw)
		bDayFormatted := t.Format("2006-01-02")

		person := Person{
			FirstName: fName,
			LastName:  lName,
			Birthday:  bDayFormatted,
			Age:       calculateAge(bDayRaw, refDate),
			Relatives: []Relative{},
		}

		// Relatives mapping: Father, Mother, Brother, Sister
		relCols := []struct {
			index int
			label string
		}{
			{3, "Father"},
			{4, "Mother"},
			{5, "Brother"},
			{6, "Sister"},
		}

		for _, relCol := range relCols {
			relName := row[relCol.index]
			if relName != "null" && relName != "" {
				rfName, rlName := parseName(relName)
				person.Relatives = append(person.Relatives, Relative{
					FirstName:    rfName,
					LastName:     rlName,
					Relationship: relCol.label,
				})
			}
		}

		people = append(people, person)
	}

	output, _ := json.MarshalIndent(people, "", "  ")
	fmt.Println(string(output))
}

File: codegen/results/v002/gemma4_31b_nothink/20260414_073749/typescript/007_beatles_interview/solution.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * Calculates age as of July 1, 2025.
 * If the person died, the age is calculated as of the date of death.
 */
function calculateAge(birthdayStr: string, diedStr: string | null): number {
    const referenceDate = new Date(2025, 6, 1); // July 1, 2025 (Month is 0-indexed)
    
    // Parse MM/DD/YYYY
    const parseDate = (s: string) => {
        const [m, d, y] = s.split('/').map(Number);
        return new Date(y, m - 1, d);
    };

    const birthDate = parseDate(birthdayStr);
    const endDate = diedStr ? parseDate(diedStr) : referenceDate;

    let age = endDate.getFullYear() - birthDate.getFullYear();
    const m = endDate.getMonth() - birthDate.getMonth();
    if (m < 0 || (m === 0 && endDate.getDate() < birthDate.getDate())) {
        age--;
    }
    return age;
}

function solve() {
    const csvPath = path.join('input', 'input.csv');
    const content = fs.readFileSync(csvPath, 'utf8');
    const lines = content.split(/\r?\n/).filter(line => line.trim() !== '');
    
    if (lines.length === 0) return;

    const headers = lines[0].split(',');
    const dataLines = lines.slice(1);

    const result = dataLines.map(line => {
        const values = line.split(',');
        const row: Record<string, string> = {};
        headers.forEach((h, i) => {
            row[h] = values[i];
        });

        const fullNames = row['Name'].split(' ');
        const firstName = fullNames[0];
        const lastName = fullNames[fullNames.length - 1];
        
        const birthday = row['Birthday'];
        const died = row['Died'] === 'null' ? null : row['Died'];
        
        const relatives = [];
        const relativeTypes = ['Father', 'Mother', 'Brother', 'Sister'];
        
        relativeTypes.forEach(type => {
            const val = row[type];
            if (val && val !== 'null') {
                const parts = val.split(' ');
                relatives.push({
                    FirstName: parts[0],
                    LastName: parts[parts.length - 1],
                    Relationship: type
                });
            }
        });

        return {
            FirstName: firstName,
            LastName: lastName,
            Birthday: `${birthday.split('/')[2]}-${birthday.split('/')[0].padStart(2, '0')}-${birthday.split('/')[1].padStart(2, '0')}`,
            Age: calculateAge(birthday, died),
            Relatives: relatives
        };
    });

    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
}

solve();